File: leetcode-solution/leetcode90-lintcode18-subset_II/src/solution.java
import java.util.ArrayList;
import java.util.Arrays;

/**
 * Created by Felicity on 2017/3/22.
 */
public class solution {
    /**
     * @param nums: A set of numbers.
     * @return: A list of lists. All valid subsets.
     */
    public ArrayList<ArrayList<Integer>> subsetsWithDup(int[] nums) {
        // write your code here
        ArrayList<ArrayList<Integer>> results = new ArrayList<>();
        if (nums == null) {
            return results;
        }

        Arrays.sort(nums);
        helper(results, new ArrayList<Integer>(), nums, 0);
        return results;
    }
    private void helper(ArrayList<ArrayList<Integer>> results,
                        ArrayList<Integer> subset,
                        int[] nums,
                        int startIndex) {
        results.add(new ArrayList<Integer>(subset));
        for (int i = startIndex; i < nums.length; i++) {
            /*
            startIndex指的是nums的元素下标，以nums为[1,1',1'']为例，如果
            startIndex是0，那么subset是[1]，如果startIndex是1，那么subset是[1,1']。
             */
            if (i != 0 && nums[i] == nums[i - 1] && i > startIndex) {
                continue;
            }
            subset.add(nums[i]);
            helper(results, subset, nums, i + 1);
            subset.remove(subset.size() - 1);
        }
    }
}

File: leetcode-solution/leetcode90-lintcode18-subset_II/Solution.py
class Solution:
    """
    @param: nums: A set of numbers.
    @return: A list of lists. All valid subsets.
    """
    # 2**len(nums)是包含nums里的所有元素的子集个数
    # >> 是右位移运算符
    def subsetsWithDup(self, nums):
        # write your code here
        res = [[]]
        nums.sort()
        for i in range(len(nums)):
            # if S[i] is same to S[i - 1], then it needn't to be added to all of the subset, just add it to the last l subsets which are created by adding S[i - 1]
            if i == 0 or nums[i] != nums[i - 1]:
                # "l" was the number of subsets created by the last s[j] where s[j] != s[j-1].
                l = len(res)
            for j in range(len(res) - l, len(res)):
                res.append(res[j] + [nums[i]])
                #不明白的可动笔画一画
        return res
